calico_containers/pycalico/ipam.py

```diff
@@ -15,78 +15,472 @@
+            for block_cidr in pool.cidr.subnet(BLOCK_PREFIXLEN[version]):
+                block_id = str(block_cidr)
+                _log.debug("Checking if block %s is free.", block_id)
+                key = _datastore_key(block_cidr)
+                try:
+                    _ = self.etcd_client.read(key)
+                except EtcdKeyNotFound:
+                    _log.debug("Found block %s free.", block_id)
+                    try:
+                        self._claim_block_affinity(host, block_cidr)
+                    except KeyError:
+                        # Failed to claim the block because some other host
+                        # has it.
+                        _log.debug("Failed to claim block %s", block_cidr)
+                        continue
+                    # Success!
+                    return block_cidr
+        raise NoFreeBlocksError()
+
+    def _claim_block_affinity(self, host, block_cidr):
+        """
+        Claim a block we think is free.
+        """
+        block_id = str(block_cidr)
+        path = IPAM_HOST_AFFINITY_PATH % {"hostname": host,
+                                          "version": block_cidr.version}
+        key = path + block_id.replace("/", "-")
+        self.etcd_client.write(key, "")
+
+        # Create the block.
+        block = AllocationBlock(block_cidr, host)
+        try:
+            self._compare_and_swap_block(block)
+        except CASError:
+            # Block exists.  Read it back to find out its host affinity
+            block = self._read_block(block_cidr)
+            if block.host_affinity == host:
+                # Block is now claimed by us.  Some other process on this host
+                # must have claimed it.
+                _log.debug("Block %s already claimed by us. Success.",
+                           block_cidr)
+                return
+
+            # Some other host beat us to claiming this block.  Clean up.
+            self.etcd_client.delete(key)
+
+            # Throw a key error to let the caller know the block wasn't free
+            # after all.
+
+            raise KeyError(block_id)
+        # successfully created the block.  Done.
+        return
+
+
+class CASError(Exception):
+    """
+    Compare-and-swap atomic update failed.
+    """
+    pass
+
+
+class NoFreeBlocksError(Exception):
+    """
+    Tried to get a new block but there are none available.
+    """
+    pass
+
 
-        while True:
-            assigned_addresses = self.etcd.get_assigned_addresses(pool)
+def _datastore_key(block_cidr):
+    """
+    Translate a block_id into a datastore key.
+    :param block_cidr: IPNetwork representing the block
+    :return: etcd key as string.
+    """
+    path = IPAM_BLOCK_PATH % {'version': block_cidr.version}
+    return path + str(block_cidr).replace("/", "-")
+
+
+class IPAMClient(BlockReaderWriter):
+
+    def auto_assign_ips(self, num_v4, num_v6, primary_key, attributes,
+                        pool=(None, None)):
```

Comment:
num_v4 and num_v6 are separate parms.  Any reason why pool is a tuple for v4 and v6?